File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{air_quality_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Air Quality Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let aq = AirQualityBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let all_values_receiver = aq.get_all_values_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `aq` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for all_values in all_values_receiver {
            println!("IAQ Index: {}", all_values.iaq_index);

            if all_values.iaq_index_accuracy == AIR_QUALITY_BRICKLET_ACCURACY_UNRELIABLE {
                println!("IAQ Index Accuracy: Unreliable");
            } else if all_values.iaq_index_accuracy == AIR_QUALITY_BRICKLET_ACCURACY_LOW {
                println!("IAQ Index Accuracy: Low");
            } else if all_values.iaq_index_accuracy == AIR_QUALITY_BRICKLET_ACCURACY_MEDIUM {
                println!("IAQ Index Accuracy: Medium");
            } else if all_values.iaq_index_accuracy == AIR_QUALITY_BRICKLET_ACCURACY_HIGH {
                println!("IAQ Index Accuracy: High");
            }

            println!("Temperature: {} °C", all_values.temperature as f32 / 100.0);
            println!("Humidity: {} %RH", all_values.humidity as f32 / 100.0);
            println!("Air Pressure: {} hPa", all_values.air_pressure as f32 / 100.0);
            println!();
        }
    });

    // Set period for all values callback to 1s (1000ms).
    aq.set_all_values_callback_configuration(1000, false);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/php/ExampleCallback.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletAirQuality.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletAirQuality;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your Air Quality Bricklet

// Callback function for all values callback
function cb_allValues($iaq_index, $iaq_index_accuracy, $temperature, $humidity,
                      $air_pressure)
{
    echo "IAQ Index: $iaq_index\n";

    if ($iaq_index_accuracy == BrickletAirQuality::ACCURACY_UNRELIABLE) {
        echo "IAQ Index Accuracy: Unreliable\n";
    } elseif ($iaq_index_accuracy == BrickletAirQuality::ACCURACY_LOW) {
        echo "IAQ Index Accuracy: Low\n";
    } elseif ($iaq_index_accuracy == BrickletAirQuality::ACCURACY_MEDIUM) {
        echo "IAQ Index Accuracy: Medium\n";
    } elseif ($iaq_index_accuracy == BrickletAirQuality::ACCURACY_HIGH) {
        echo "IAQ Index Accuracy: High\n";
    }

    echo "Temperature: " . $temperature/100.0 . " °C\n";
    echo "Humidity: " . $humidity/100.0 . " %RH\n";
    echo "Air Pressure: " . $air_pressure/100.0 . " hPa\n";
    echo "\n";
}

$ipcon = new IPConnection(); // Create IP connection
$aq = new BrickletAirQuality(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Register all values callback to function cb_allValues
$aq->registerCallback(BrickletAirQuality::CALLBACK_ALL_VALUES, 'cb_allValues');

// Set period for all values callback to 1s (1000ms)
$aq->setAllValuesCallbackConfiguration(1000, FALSE);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
